src/ray/raylet/scheduling/fair_scheduling_queue.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+/* #include <tuple> */
+#include "ray/common/task/task.h"
+#include "ray/rpc/node_manager/node_manager_server.h"
+
+namespace ray {
+namespace raylet {
+
+/// Work represents all the information needed to make a scheduling decision.
+/// This includes the task, the information we need to communicate to
+/// dispatch/spillback and the callback to trigger it.
+using Work = std::tuple<Task, rpc::RequestWorkerLeaseReply *, std::function<void(void)>>;
+
+namespace internal {
+
+struct SchedulingPriorityComparator : public std::less<SchedulingClass> {
+ public:
+  explicit SchedulingPriorityComparator(
+      std::unordered_map<SchedulingClass, uint64_t> &priorities);
+
+  bool operator()(const SchedulingClass &lhs, const SchedulingClass &rhs) const;
+
+ private:
+  /// A reference to FairSchedulingQueue::active_works_
+  std::unordered_map<SchedulingClass, uint64_t> &priorities_;
+};
+
+using WorkQueueType =
+    std::map<SchedulingClass, std::deque<Work>, internal::SchedulingPriorityComparator>;
+}  // namespace internal
+
+/* using WorkQueueIterator = std::iterator<std::forward_iterator_tag, const
+ * std::pair<const SchedulingClass, std::deque<Work>>>; */
+using WorkQueueIterator = internal::WorkQueueType::iterator;
+using ConstWorkQueueIterator = internal::WorkQueueType::const_iterator;
+
+class FairSchedulingQueue {
```

Comment:
Please write some comments to explain what's this queue for. Maybe give some examples to explain the behavior under the hood.